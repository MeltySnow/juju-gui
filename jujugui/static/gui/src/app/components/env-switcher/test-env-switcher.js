/* Copyright (C) 2017 Canonical Ltd. */
'use strict';

const React = require('react');
const enzyme = require('enzyme');

const EnvSwitcher = require('./env-switcher');

const modelResponse = require('@canonical/jujulib/tests/data/modelmanager-response');

describe('EnvSwitcher', function() {
  const renderComponent = (options = {}) => enzyme.shallow(
    <EnvSwitcher.WrappedComponent
      acl={{}}
      addNotification={options.addNotification || sinon.stub()}
      changeState={options.changeState || sinon.stub()}
      environmentName={options.environmentName || 'MyEnv'}
      listModelSummaries={options.listModelSummaries || sinon.stub()}
      modelCommitted={
        options.modelCommitted === undefined ? false : options.modelCommitted}
      setModelName={options.setModelName || sinon.stub()}
      switchModel={options.switchModel || sinon.stub()}
      user={{}} />
  );

  it('renders the closed switcher component', () => {
    const wrapper = renderComponent();
    expect(wrapper).toMatchSnapshot();
  });

  it('should not have an editable name when the model is committed', () => {
    const wrapper = renderComponent({modelCommitted: true});
    expect(wrapper.find('.env-switcher__name')).toMatchSnapshot();
  });

  it('defaults to "untitled-model"', () => {
    // Shallow rendering a wrapped component doesn't appear to apply the
    // defaultProps. Because other tests test that the environmentName is
    // correctly applied. This check to see that a default is defined should
    // suffice.
    assert.deepEqual(EnvSwitcher.WrappedComponent.defaultProps, {
      environmentName: 'untitled-model'
    });
  });

  it('can change the model name', () => {
    const setModelName = sinon.stub();
    const wrapper = renderComponent({setModelName});
    const instance = wrapper.instance();
    instance.refs = {
      name: {
        innerText: 'new-name'
      }
    };
    wrapper.find('.env-switcher__name').props().onBlur();
    assert.equal(setModelName.callCount, 1);
    assert.equal(setModelName.args[0][0], 'new-name');
  });

  it('does not change the model name if there is an error', () => {
    const setModelName = sinon.stub();
    const wrapper = renderComponent({setModelName});
    const instance = wrapper.instance();
    instance.refs = {
      name: {
        innerText: '-'
      }
    };
    wrapper.find('.env-switcher__name').props().onBlur();
    assert.equal(setModelName.callCount, 0);
  });

  it('can render when there is an error', () => {
    const wrapper = renderComponent();
    const instance = wrapper.instance();
    instance.refs = {
      name: {
        innerText: '-'
      }
    };
    wrapper.find('.env-switcher__name').props().onBlur();
    wrapper.update();
    assert.strictEqual(wrapper.prop('className').includes('env-switcher--error'), true);
  });

  it('opens the list and fetches models on click', () => {
    const listModelSummaries =
      sinon
        .stub()
        .callsArgWith(1, null, modelResponse.listModelSummaries.response);
    const wrapper = renderComponent({listModelSummaries});
    assert.equal(wrapper.find('EnvList').length, 0);
    wrapper.find('.env-switcher__chevron').simulate('click', {
      preventDefault: () => null
    });
    const envList = wrapper.find('EnvList');
    assert.equal(envList.length, 1);
    assert.equal(envList.prop('environmentName'), 'MyEnv');
    expect(envList.prop('envs')).toMatchSnapshot();
  });

  it('can call to switch models', () => {
    // To switch environments you click on an environment list item in a sub
    // component so here we're just going to call the method that gets
    // passed down.
    const listModelSummaries =
      sinon
        .stub()
        .callsArgWith(1, null, modelResponse.listModelSummaries.response);
    const switchModel = sinon.stub();
    const wrapper = renderComponent({listModelSummaries, switchModel});
    const instance = wrapper.instance();
    const model = {
      id: 'abc123',
      name: 'Tardis',
      owner: 'The Dr.'
    };
    instance.handleModelClick(model);
    assert.equal(switchModel.callCount, 1);
    assert.deepEqual(switchModel.args[0], [model]);
  });

  it('handles errors when getting models', function() {
    const addNotification = sinon.stub();
    const listModelSummaries = sinon.stub().callsArgWith(1, 'Uh oh!', null);
    renderComponent({listModelSummaries, addNotification});
    assert.equal(addNotification.callCount, 1);
    assert.deepEqual(addNotification.args[0][0], {
      title: 'unable to retrieve model list',
      message: 'unable to retrieve model list: Uh oh!',
      level: 'error'
    });
  });
});
