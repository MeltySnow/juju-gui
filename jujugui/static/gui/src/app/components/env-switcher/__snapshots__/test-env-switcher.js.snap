// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvSwitcher opens the list and fetches models on click 1`] = `
Array [
  Object {
    "result": Object {
      "agent-version": "2.4.5",
      "cloud-credential-tag": "cloudcred-google_thedr@external_admin",
      "cloud-region": "us-central1",
      "cloud-tag": "cloud-google",
      "controller-uuid": "a030379b-940f-4760-9fcf-3062b41a04e7",
      "counts": Array [
        Object {
          "count": 0,
          "entity": "machines",
        },
        Object {
          "count": 0,
          "entity": "cores",
        },
      ],
      "default-series": "bionic",
      "last-connection": null,
      "life": "alive",
      "name": "group-test",
      "owner-tag": "user-thedr@external",
      "provider-type": "gce",
      "sla": null,
      "status": Object {
        "info": "",
        "since": "2018-11-07T05:02:42.436Z",
        "status": "available",
      },
      "type": "",
      "user-access": "admin",
      "uuid": "57650e3c-815f-4540-89df-81fd5d70b7ef",
    },
  },
  Object {
    "result": Object {
      "agent-version": "2.4.5",
      "cloud-credential-tag": "cloudcred-google_thedr@external_admin",
      "cloud-region": "us-central1",
      "cloud-tag": "cloud-google",
      "controller-uuid": "a030379b-940f-4760-9fcf-3062b41a04e7",
      "counts": Array [
        Object {
          "count": 0,
          "entity": "machines",
        },
        Object {
          "count": 0,
          "entity": "cores",
        },
      ],
      "default-series": "xenial",
      "last-connection": null,
      "life": "alive",
      "name": "mymodel",
      "owner-tag": "user-thedr@external",
      "provider-type": "gce",
      "sla": null,
      "status": Object {
        "info": "",
        "since": "2018-11-07T05:02:41.712Z",
        "status": "available",
      },
      "type": "",
      "user-access": "admin",
      "uuid": "9ed7302e-6f44-4c87-8ac5-68bb38efb843",
    },
  },
  Object {
    "result": Object {
      "agent-version": "2.4.5",
      "cloud-credential-tag": "cloudcred-google_thedr@external_juju",
      "cloud-region": "us-central1",
      "cloud-tag": "cloud-google",
      "controller-uuid": "a030379b-940f-4760-9fcf-3022b41a04e7",
      "counts": Array [
        Object {
          "count": 0,
          "entity": "machines",
        },
        Object {
          "count": 0,
          "entity": "cores",
        },
      ],
      "default-series": "xenial",
      "last-connection": null,
      "life": "alive",
      "name": "single-cred",
      "owner-tag": "user-thedr@external",
      "provider-type": "gce",
      "sla": null,
      "status": Object {
        "info": "",
        "since": "2018-11-07T05:02:41.712Z",
        "status": "available",
      },
      "type": "",
      "user-access": "admin",
      "uuid": "9ed7302e-6f22-4c87-8ac5-68bb38efb843",
    },
  },
  Object {
    "result": Object {
      "agent-version": "2.4.3",
      "cloud-credential-tag": "cloudcred-google_radman@external_google-radmanbot",
      "cloud-region": "us-east1",
      "cloud-tag": "cloud-google",
      "controller-uuid": "a030379b-940f-4760-9fcf-3062b41a04e7",
      "counts": Array [
        Object {
          "count": 13,
          "entity": "machines",
        },
        Object {
          "count": 13,
          "entity": "cores",
        },
      ],
      "default-series": "xenial",
      "last-connection": null,
      "life": "alive",
      "name": "kpi",
      "owner-tag": "user-radman@external",
      "provider-type": "gce",
      "sla": null,
      "status": Object {
        "info": "",
        "since": "2018-10-29T08:12:59.512Z",
        "status": "available",
      },
      "type": "",
      "user-access": "read",
      "uuid": "06b11eb5-abd6-48f3-8910-b54cf5905e60",
    },
  },
  Object {
    "result": Object {
      "agent-version": "2.2.6",
      "cloud-credential-tag": "cloudcred-google_bigboss@external_gce",
      "cloud-region": "us-east1",
      "cloud-tag": "cloud-google",
      "controller-uuid": "a030379b-940f-4760-9fcf-3062b41a04e7",
      "counts": Array [
        Object {
          "count": 3,
          "entity": "machines",
        },
        Object {
          "count": 5,
          "entity": "cores",
        },
      ],
      "default-series": "xenial",
      "last-connection": null,
      "life": "alive",
      "name": "ci",
      "owner-tag": "user-bigboss@external",
      "provider-type": "gce",
      "sla": null,
      "status": Object {
        "info": "",
        "since": "2018-10-29T08:12:58.792Z",
        "status": "available",
      },
      "type": "",
      "user-access": "admin",
      "uuid": "4e8a5213-c34a-4ac9-83fa-1175165467d7",
    },
  },
]
`;

exports[`EnvSwitcher renders the closed switcher component 1`] = `
<div
  aria-label="Model switcher"
  className="env-switcher"
  onClick={null}
  role="navigation"
  tabIndex="0"
>
  <div
    className="env-switcher__toggle editable"
  >
    <div>
      <span
        className="env-switcher__name"
        contentEditable={true}
        dangerouslySetInnerHTML={
          Object {
            "__html": "MyEnv",
          }
        }
        onBlur={[Function]}
        onFocus={[Function]}
      />
      <div
        className="env-switcher__name-error"
      >
        The model name must only contain lowercase letters, numbers, and hyphens. It must not start or end with a hyphen.
      </div>
    </div>
    <div
      aria-controls="environmentSwitcherMenu"
      aria-expanded="false"
      aria-haspopup="true"
      aria-owns="environmentSwitcherMenu"
      className="env-switcher__chevron"
      id="environmentSwitcherToggle"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    >
      <SvgIcon
        name="chevron_down_16"
        size="16"
      />
    </div>
  </div>
</div>
`;

exports[`EnvSwitcher should not have an editable name when the model is committed 1`] = `
<span
  className="env-switcher__name"
>
  MyEnv
</span>
`;
